packages/swing-store/src/transcriptStore.js

```diff
@@ -0,0 +1,383 @@
+   * @param {number} startPos
+   * @returns {AsyncIterable<Uint8Array>}
+   */
+  function getTranscript(vatID, startPos) {
+    async function* fetcher() {
+      for (const entry of readTranscript(vatID, startPos)) {
+        const item = typeof entry === 'string' ? Buffer.from(entry) : entry;
+        yield item;
+        yield nl;
+      }
+    }
+    return fetcher();
+  }
+
+  function updateTranscriptHash(priorHash, item) {
+    const hasher = createSHA256();
+    hasher.add(priorHash);
+    hasher.add(item);
+    return hasher.finish();
+  }
+
+  const sqlWriteTranscriptItem = db.prepare(`
+    INSERT INTO transcriptItems (vatID, item, position)
+    VALUES (?, ?, ?)
+    ON CONFLICT(vatID, position) DO UPDATE SET item = ?
+  `);
+
+  const sqlUpdateTranscriptSpan = db.prepare(`
+    UPDATE transcriptSpans
+    SET endPos = ?, hash = ?, size = ?
+    WHERE vatID = ? AND current = 1
+  `);
+
+  /**
+   * @param {string} vatID
+   * @param {string} item
+   */
+  const writeTranscriptItem = (vatID, item) => {
+    ensureTxn();
+    const { endPos, size, hash } = getCurrentTranscriptBounds(vatID);
+    !transcriptStatus.get(vatID) ||
+      Fail`can't write transcript ${q(vatID)} because it's already in use`;
+    item = `${item}\n`;
+    sqlWriteTranscriptItem.run(vatID, item, endPos, item);
+    sqlUpdateTranscriptSpan.run(
+      endPos + 1,
+      updateTranscriptHash(hash, item),
+      size + item.length,
+      vatID,
+    );
+  };
+
+  /**
+   * @param {string} name
+   * @param {SwingStoreExporter} exporter
+   * @param {Map} metadata
+   * @returns {Promise<void>}
+   */
+  async function importTranscript(name, exporter, metadata) {
+    const info = JSON.parse(metadata.get(name));
+    const parts = name.split('.');
+    const [tag, type, vatID, rawStartPos] = parts;
+    assert(
+      parts.length === 4 && tag === 'export' && type === 'transcript',
+      `expected transcript name of the form 'export.transcript.{vatID}.{startPos}', saw '${name}'`,
+    );
+    assert(
+      info.vatID === vatID,
+      `transcript name says vatID ${vatID}, metadata says ${info.vatID}`,
+    );
+    const startPos = Number(rawStartPos);
+    assert(
+      info.startPos === startPos,
+      `transcript name says startPos ${startPos}, metadata says ${info.startPos}`,
+    );
+    const artifactChunks = exporter.getArtifact(name);
+    const inStream = Readable.from(artifactChunks);
+    const lineTransform = new ReadlineTransform();
+    const lineStream = inStream.pipe(lineTransform);
+    let hash = initialHash;
+    let pos = startPos;
+    let totalSize = 0;
+    for await (const line of lineStream) {
+      const item = `${line}\n`;
+      sqlWriteTranscriptItem.run(vatID, item, pos, item);
+      hash = updateTranscriptHash(hash, item);
+      pos += 1;
+      totalSize += item.length;
+    }
+    assert(
+      info.size === totalSize,
+      `transcript ${name} size is ${totalSize}, metadata says ${info.size}`,
+    );
```

Comment:
The size check is just a bit of internal sanity checking that the metadata and the artifact are in sync with each other.  It's not strictly necessary since if the size is off then the hash will be too and so I vacillated about including it, but came down on the side of suspenders AND belt checking it since it was cheap and easy to do.

The UTF encoding thing is a distraction since if the UTF encoding varies the hash will vary and we'll be hosed.

Reply:
I don't see why UTF encoding would vary the hash, it's just the encoding used during I/O. What it means though is that this size is not guaranteed to match the artifact size (unless all characters fit in the latin 7-bit space), so the sanity check can only happen after decoding, not simply looking at the artifact size.